src/systems/filecoin_blockchain/chainsync/_index.md

```diff
@@ -1,58 +1,490 @@
+  - Nodes DO NOT need to but MAY sync earlier `StateTrees` than `LastTrustedCheckpoint` as well.
+- Pseudocode 1: a basic version of `SYNC_CHECKPOINT`:
+    ```go
+    func (c *ChainSync) SyncCheckpoint() {
+        while !c.HasCompleteStateTreeFor(c.LastTrustedCheckpoint) {
+            selector := ipldselector.SelectAll(c.LastTrustedCheckpoint)
+            c.Graphsync.Pull(c.Peers, sel, c.IpldStore)
+            // Pull SHOULD NOT pull what c.IpldStore already has (check first)
+            // Pull SHOULD pull from different peers simultaneously
+            // Pull SHOULD be efficient (try different parts of the tree from many peers)
+            // Graphsync implementations may not offer these features. These features
+            // can be implemented on top of a graphsync that only pulls from a single
+            // peer and does not check local store first.
+        }
+        c.ChainCatchup() // on to CHAIN_CATCHUP
+    }
+    ```
+- **security conditions to transition out:**
+  - `StateTree` for `LastTrustedCheckpoint` MUST be stored locally and verified (hashing is enough)
+- **transitions out:**
+  - once node receives and verifies complete `StateTree` for `LastTrustedCheckpoint`: move to `CHAIN_CATCHUP`
+
+## ChainSync FSM: `CHAIN_CATCHUP`
+
+- While in this state:
+  - `ChainSync` is well-bootstrapped, and has an initial **trusted** `StateTree` to start from.
+  - `ChainSync` is receiving latest `Blocks` from `BlockPubsub`
+  - `ChainSync` starts fetching and validating blocks (see _Block Fetching and Validation_ above).
+  - `ChainSync` has unvalidated blocks between `ChainSync.FinalityTipset` and `ChainSync.TargetHeads`
+- **Chain State and Finality**:
+  - In this state, the **chain MUST NOT advance** beyond whatever the node already has:
+    - `FinalityTipset` does not change.
+    - No new blocks are reported to consumers/users of `ChainSync` yet.
+    - The chain state provided is the available `Blocks` and `StateTree` for all available epochs,
+      specially the `FinalityTipset`.
+    - finality must not move forward here because there are serious attack vectors where a node can be forced to end up on the wrong fork if finality advances before validation is complete up to the block production fringe.
+  - Validation must advance, all the way to the block production fringe:
+    - Validate the whole chain, from `FinalityTipset` to `BestTargetHead`
+    - The node can reach `BestTargetHead` only to find out it was invalid, then has to update `BestTargetHead` with next best one, and sync to it
+      (without having advanced `FinalityTipset` yet, as otherwise we may end up on the wrong fork)
+- **security conditions to transition out:**
+  - Gaps between `ChainSync.FinalityTipset ... ChainSync.BestTargetHead` have been closed:
+    - All `Blocks` and their content MUST be fetched, stored, linked, and validated locally.
```

Comment:
I think this condition may be too strict. That is we need only have those blocks in the heaviest chain have been f/s/l/v locally. We are going depth first i believe to drive this (though obviously some breadth will be covered so doing). As soon as the heaviest "heard about" chain is validated, you can transition out. You needn't validate all other chains in order to move forward i believe.

Reply:
> "Gaps between `ChainSync.FinalityTipset ... ChainSync.BestTargetHead` have been closed:"

this means only the chain to `BestTargetHead`, so i think it covers your concern.

note that if that one turns out to be invalid, need to keep validating towards the next best one, before moving to `CHAIN_FOLLOW`

Reply:
Yes. You're completely right. I focused on the "all blocks" but it is appropriately scoped. Thank you